### PR TITLE
Tested Ubuntu 16.04 version

### PR DESCRIPTION
**Pull request by purohit**
Hello,

As promised here's a pull request for CellProfiler on Ubuntu 16.04 that's automatically tested against a real CP run. I generated the gold files by running the human cell example from your website once, and uploaded those to a public S3 location, which we're happy to provide.

A few changes:

1. I removed the warez-style header
2. We use s6-overlay to prevent the Linux zombie reaping problem instead of phusion-base-image since it's much more lightweight. This occurs when components of CellProfiler fail and don't get cleaned up properly (read more about it here: https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/).
3. Travis-CI is set-up, but you'll need to enable it on your repo (it's free for open source projects) and correspondingly change the image URL.
4. Added instructions on running the Docker container in GUI mode in case anyone's interested.

Let me know what you think!

Thanks.

### CODE REVIEW

**Comment by mcquin** [comment]
Hi @purohit. Nice work! I'll get to reviewing this ASAP.

In the meantime, could you please resolve the merge conflicts?

File: README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[![Build
+Status](https://travis-ci.com/recursionpharma/cellprofiler-docker.svg?token=GrWRwBK4pgTygz4Nzroz&branch=master)](https://travis-ci.com/recursionpharma/cellprofiler-docker)
```
Comment (mcquin) [inline README.md]: This should link to the base repository:

```
[![Build Status](https://img.shields.io/travis/CellProfiler/docker/master.svg)](https://travis-ci.org/CellProfiler/docker)
```

**Comment by mcquin** [comment]
Travis is enabled. It should run on the next commit.

**Comment by purohit** [comment]
@mcquin Thanks for taking a look. I've resolved merge conflicts and used the CellProfiler/docker Travis icon, and the build is passing!

**Comment by mcquin** [comment]
@purohit When we release CP 3.0, would you be willing to update the gold files?

**Comment by purohit** [comment]
Yep!

2017-05-11 14:44 GMT+02:00 Claire McQuin <notifications@github.com>:

File: Dockerfile [hunk 2]
```diff
@@ -1,81 +1,36 @@
...
+        python-numpy       \
+        python-vigra       \
+        python-wxgtk3.0    \
+        python-zmq
 WORKDIR /usr/local/src
 ARG VERSION=stable
 RUN git clone -b $VERSION https://github.com/CellProfiler/CellProfiler.git
 WORKDIR /usr/local/src/CellProfiler
+# Install CellProfiler at a specific version
+RUN git checkout tags/2.2.0
```
Comment (mcquin) [inline Dockerfile]: I understand that this will always install CellProfiler 2.2.0. Is this correct? Should this command make use of `VERSION`?

Comment (purohit) [reply Dockerfile]: You're right, I forgot the other PR had been merged in which used versions. I'll update.

**Comment by purohit** [comment]
@mcquin This is ready to merge, the other PR that was merged in caused some problems as the "stable" branch wasn't actually producing clean results, so I included the `ARG` option but defaulted it to `tags/2.2.0` as it was before.

**Comment by mcquin** [comment]
Hm. I'm not convinced the versioning is quite right. From the Travis log, I see: [line](https://travis-ci.org/CellProfiler/docker/builds/231188908#L274)

```
docker build -t cellprofiler:latest .
```

Then, it appears to checkout CP 2.2.0: [lines](https://travis-ci.org/CellProfiler/docker/builds/231188908#L3125-L3131)

```
Step 6 : ARG VERSION=tags/2.2.0
 ---> Running in db5dd2dbe042
 ---> 77ffdd3aabcd
Removing intermediate container db5dd2dbe042
Step 7 : RUN git checkout $VERSION
 ---> Running in 4af0addf6307
Note: checking out 'tags/2.2.0'.
```

It should grab master instead?

File: Dockerfile [hunk 3]
```diff
@@ -1,81 +1,34 @@
...
+        cython             \
+        git                \
+        libmysqlclient-dev \
+        libhdf5-dev        \
+        libxml2-dev        \
+        libxslt1-dev       \
+        openjdk-8-jdk      \
+        python-dev         \
+        python-pip         \
+        python-h5py        \
```
Comment (0x00b1) [inline Dockerfile:15]: Let’s use the versions of H5Py, matplotlib, NumPy, and SciPy available from PyPI.

File: Dockerfile [hunk 4]
```diff
@@ -1,81 +1,34 @@
...
+        python-matplotlib  \
+        python-mysqldb     \
+        python-scipy       \
+        python-numpy       \
+        python-vigra       \
```
Comment (0x00b1) [inline Dockerfile:20]: `python-vigra` was needed by CellProfiler’s Ilastik support. It can be safely removed.

File: Dockerfile [hunk 5]
```diff
@@ -1,81 +1,34 @@
...
-RUN git clone -b $VERSION https://github.com/CellProfiler/CellProfiler.git
+# Install CellProfiler
+RUN git clone https://github.com/CellProfiler/CellProfiler.git
 WORKDIR /usr/local/src/CellProfiler
-RUN pip install                                                     \
-  --editable                                                        \
-    .
-ENTRYPOINT ["cellprofiler", "--run", "--run-headless"]
-CMD ["--help"]
+ARG VERSION=tags/2.2.0
```
Comment (0x00b1) [inline Dockerfile:27]: We should start thinking about 3.0.0.

File: Dockerfile [hunk 6]
```diff
@@ -1,81 +1,34 @@
-#                                        Ü       ÜÜÜÜÜÜÜÜÜÜÜ
```
Comment (0x00b1) [inline Dockerfile:1]: You say “warez,” I say “demoscene.”

Comment (purohit) [reply Dockerfile:1]: :laughing:

**Comment by purohit** [comment]
@mcquin It is intentionally grabbing the tagged 2.2.0 version since that's the latest 2.x version I know to work with the tests -- when I use master or stable, it breaks. The actual docker tag is just for convenience in the Makefile and can be user specified (i.e. `latest` by Docker convention).

I would recommend we merge these changes and then make @0x00b1's future improvements in other pull requests. I'm happy to help with them as time allows, or others can chip in, but don't want to prolong this pull request beyond the scope of "let's get this working and tested". Once we have a CI-tested docker base, it will be much easier for us to safely make those changes.

File: Dockerfile (hunk 5 above)
Comment (purohit) [reply Dockerfile:27]: Agreed!

**Comment by purohit** [comment]
On second thought regarding `ARG`, it may be a bad idea to allow `ARG` in the Dockerfile specifying the branch to checkout. The code significantly differs based on the branch, and changing it will erase the guarantee that the tests here provide; namely, that the container works correctly.

If people want to customize the Dockerfile, that's fine, but it doesn't need to be canon.

What do you think?

**Comment by mcquin** [comment]
> The code significantly differs based on the branch, and changing it will erase the guarantee that the tests here provide; namely, that the container works correctly.

Is there another test for correctness that is version independent? For example, I can run the ExampleFlyPipeline and observe certain image and CSV files were created, and these are non-empty. CellProfiler's unit tests ensure correctness of data. Perhaps the container tests could include running the unit tests as well (although, this takes a few minutes).

**Comment by purohit** [comment]
> Is there another test for correctness that is version independent? For example, I can run the ExampleFlyPipeline and observe certain image and CSV files were created, and these are non-empty. 

I suppose non-empty would indicate that something is being done (good sign), although I was looking for a guarantee that when someone changes the Dockerfile for a pull request, that the output of CellProfiler is the same (so we know nothing got messed up). But I guess using `build-arg` with defaults doesn't prevent this.

> Perhaps the container tests could include running the unit tests as well (although, this takes a few minutes).

That could work, though I'd prefer a shorter build time and let CellProfiler/CellProfiler handle the unit tests.

Anyway, let me know if there's anything left preventing this changed from being merged.

**Pull request merged**